compiler-core/src/type_.rs

```diff
@@ -134,6 +134,33 @@ impl Type {
         }
     }
 
+    /// Gets the types inside of a tuple. Returns `None` if the type is not a tuple.
+    pub fn tuple_types(&self) -> Option<Vec<Arc<Self>>> {
+        match self {
+            Self::Tuple { elems } => Some(elems.clone()),
+            _ => None,
+        }
+    }
+
+    /// Gets the argument types for a type constructor. Returns `None` if the type
+    /// does not lead to a type constructor.
+    pub fn constructor_types(&self) -> Option<Vec<Arc<Self>>> {
+        match self {
+            Self::Named { args, .. } => Some(args.clone()),
+            Self::Var { type_, .. } => type_.borrow().constructor_types(),
+            _ => None,
+        }
+    }
+
+    /// Gets the inner type following any var links. Returns a pointer to self
+    /// if the type is not a var.
+    pub fn inner_type(&self) -> Option<Arc<Self>> {
+        match self {
+            Self::Var { type_, .. } => type_.borrow().inner_type(),
+            _ => Some(Arc::new(self.clone())),
+        }
+    }
```

Comment:
I think this is called collapse_type_links elsewhere in the compiler and I think it doesn't clone the type if it is not a link.

Reply:
I spent more time than i want to admit trying to figure out how to do the recursion on the inner `Arc<RefCell<TypeVar>>` and it looks like switching to `collapse_links` actually solves that much nicer than the current implementation. Thanks for pointing this out